### PR TITLE
Migrate to multiversion 0.7, better matmul example

### PR DESCRIPTION
**Pull request by HadrienG2**
So, in preparation for #8 I wanted to move to multiversion v0.7 because its API is so much simpler. But one problem was the matmul example, which made use of advanced multiversion v0.6 tricks that are not possible with v0.7.

Since that example was not a very good SIMD matmul anyway, I thought I would rewrite it into something that's AFAIK optimal modulo alignment issues. But maybe that's now a little too complex for example code though. Tell me what you think!

### CODE REVIEW

**Comment by HadrienG2** [comment]
Ping @vorner ?

**Comment by vorner** [comment]
Sorry. I'm having it open in one of my browser tabs and trying to get around to it. From a quick glance, I'd have preferred something more lightweight as an example… but I'll give it a proper read soon.

**Comment by HadrienG2** [comment]
I was wondering if it wouldn't make sense to have a dot product example (which is much simpler to optimize and you already have it as a test) and link back to that with an introductory comment like "(Efficient) SIMD matrix multiplication is quite a bit harder than SIMD dot product, if you're just getting started with SIMD or this library you may want to check out the dot product example first".

The reason why I think it's important to do matmul right even if it makes the code more complicated is that this library is fundamentally about performance optimization, and targeted at people familiar with the craft. So if we don't show something close to the 32x speedup theoretically expected from AVX + ILP + FMA, it may make the library a harder sell to people familiar with hardware specs who know what to expect.

**Comment by vorner** [comment]
> The reason why I think it's important to do matmul right even if it makes the code more complicated

Point taken.

OK, the plan with the dot product and the introductory example should probably work.

File: examples/matrix.rs [hunk 1]
```diff
@@ -1,125 +1,160 @@
...
-#[inline]
-fn at(x: usize, y: usize) -> usize {
-    y * SIZE + x
-}
+// Number of benchmark repetitions
+const RUNS: u32 = 10_000;
+
+// FIXME: Depending on how lucky you are with memory allocator lottery, you may
+//        or may not get a vector that's properly aligned for SIMD processing.
+//        Using a Vec<V> would be better from this perspective.
```
Comment (vorner) [inline examples/matrix.rs]: I'm wondering about one thing. Allocate 7 more elements in the vector, look at the alignment of the first one and offset the start based on that. Then, if there's an assert at the start of the function, checking the proper alignment, the rest should be able to take advantage.

**Comment by HadrienG2** [comment]
I've started working on a dot product example, and there I discovered the reason for some suspicious perf numbers I was getting with matmul: the compiler optimizer does not manage to optimize `slice.vectorize()` as well as the standard `slice.chunks_exact(V::LANES).map(V::new)` equivalent.

I'll see if I can fix this...

File: examples/matrix.rs (hunk 1 above)
Comment (HadrienG2) [reply examples/matrix.rs]: In simple cases (think dot, not matmul), vectorize() could also take a hint from compiler autovectorizers and help here by splitting the loop into a (peel, body, tail) triplet, where body has an alignment and size that's a multiple of the SIMD vector size.

Comment (HadrienG2) [reply examples/matrix.rs]: ...but sadly that won't work in the case where the user specifies multiple scalar slices with different alignment. And also, it means that all scalar loops need padding, which is an undesirable ergonomics regression. So probably not a good idea.

File: examples/dot_product.rs [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+        fn $name(lhs: &Vector, rhs: &Vector) -> Scalar {
+            // Set up a single SIMD accumulator
+            let mut accumulator = V::default();
+
+            // Iterate over SIMD vectors and compute sum of products
+            for (lvec, rvec) in (&lhs.0[..], &rhs.0[..]).vectorize() {
+                if target_cfg_f!(target_feature = "fma") {
+                    accumulator = lvec.mul_add(rvec, accumulator);
+                } else {
+                    accumulator += lvec * rvec;
```
Comment (vorner) [inline examples/dot_product.rs:80]: Is it necessary, or will the „fake“ fma provided by the compiler be good enough?

**Review by vorner — COMMENTED** [review]
I'm mostly OK with how the examples look now, but should I wait with merging for some outcome of this?

> > I've started working on a dot product example, and there I discovered the reason for some suspicious perf numbers I was getting with matmul:

**Comment by HadrienG2** [comment]
Maybe let's wait a bit as I have some cool things in the pipeline, e.g. rolled out a small crate that together with the new vectorizer impl lets me have much nicer loop unrolling syntax for ILP : https://github.com/HadrienG2/slipstream/blob/78190438daae1c68873060e6ac1d745d09d1cb03/examples/dot_product.rs#L93

File: examples/dot_product.rs (hunk 2 above)
Comment (HadrienG2) [reply examples/dot_product.rs:80]: On Intel CPUs, where FMA is the same execution port as ADD and MUL, it's a factor of 2 difference when codegened right. On AMD CPUs, which have dedicated ADD and MUL ports, the picture is more nuanced.

Comment (HadrienG2) [reply examples/dot_product.rs:80]: But for the "simple" version which does not have ILP, and is thus latency-bound, using FMA is strictly a bad idea as it increases the latency with respect to mul + add. So I might want to remove it from this version to make it even simpler.

**Comment by HadrienG2** [comment]
Unfortunately the vectorize refactor turned out to be too big for my resources. I guess I should close this, unless you're interested in the subset provided by this PR?

**Comment by vorner** [comment]
Well, I haven't really touched this repository in years, so I guess I don't really have the time for it either.

**Comment by HadrienG2** [comment]
Closing then ;)

**Pull request closed without merging**